alphagov/whitehall
# Content modelling/410 schedule publishing

**Note: to make this work you will need `Sidekiq Admin` permissions - locally, this can be added to the Test User in `db/seeds.rb`, on integration this can be added to each user through SignOn.**

We want to allow users to schedule the publication of Content Blocks. This is the first attempt at introducing that in the code. As with our other features, we follow Whitehall patterns where possible/where it makes sense. This has included :
* having any actions related to the `workflow` of an Edition in a separate controller
  * @tahb and I discussed renaming the `workflow_controller` to `editions_workflow_controller` to make this more clear, will do in a future PR
* Using a Worker class to put jobs onto a Sidekiq queue.
  * our Worker is almost identical, but calls a different Service to schedule a Content Block Edition. 
  * while Whitehall have a separate `ScheduledEditionPublisher` to publish scheduled Editions, as we don't have any logic to check for scheduled editions, I've just re-used our `PublishEditionService`.
* Using a similar workflow state on the Edition, and a field called `scheduled_publication` to store the datetime in the db 
* Using the same date/time fields component  and error messages in the view.

![Screenshot 2024-08-27 at 12 34 32](https://github.com/user-attachments/assets/2b020b29-2136-4fc8-9a45-44bd9abc8977)


![Screenshot 2024-08-27 at 17 29 13](https://github.com/user-attachments/assets/ab34f044-0706-4fb9-bd1d-072fbb3d205c)

![Screenshot 2024-08-27 at 17 29 28](https://github.com/user-attachments/assets/a9321b93-a188-455c-8c65-9fea1e6a2ef7)


Sidekiq jobs can be viewed at http://whitehall-admin.dev.gov.uk/sidekiq/queues 

## steps for manual testing
* create a content block
* go to 'change' the content block details
* schedule the publication for the next minute or so
* observe job at http://whitehall-admin.dev.gov.uk/sidekiq/queues 
* once published, put content id for the block (found in embed code) and check published on Publishing API http://publishing-api.dev.gov.uk/v2/content/{content-id}

## Todo in future PRs

* if there is another scheduled edition, scheduling should not be allowed or we should supersede the existing scheduled edition
* version history/ audit trail
  * the scheduling action should appear on the Document Timeline as `scheduled` by User
  * the actual publication by Sidekiq should appear as `published` (by who?)
* Interaction design for showing the state of the Block and the publication date. 
* Interaction design to cancel or supersede a scheduled Block
* Understanding how this relates to Documents that have embedded the Block - e.g. superseding logic, notifications etc.  





~~⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️~~

~~Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.~~
